src/librustc_resolve/lib.rs

```diff
@@ -3273,6 +3274,38 @@ impl<'a> Resolver<'a> {
                     let traits = self.get_traits_containing_item(item_name, ns);
                     self.trait_map.insert(id, traits);
                 }
+
+                // check if we wrote `str::from_utf8` instead of `std::str::from_utf8`
+                let item_span = path.iter().last().map(|segment| segment.ident.span)
+                    .unwrap_or(span);
+                let mut std_path = vec![Segment::from_ident(Ident::from_str("std"))];
+                std_path.extend(path);
+                match self.resolve_qpath_anywhere(
```

Comment:
You don't need neither `resolve_qpath_anywhere`, nor `resolve_qpath` for speculative search of names in modules.
`resolve_path_without_parent_scope` is used for that (you can see some other diagnostics code using it already).

Reply:
I'm also not too happy about doing this extra work on happy path.
Perhaps we should do it only if it really looks like a built-in type (`primitive_type_table` contains it).